tests/kernel/interrupt/src/shared_irq.c

```diff
@@ -0,0 +1,337 @@
+#else
+#error "Unsupported architecture."
+#endif
+
+	arch_irq_connect_dynamic(irq_line_1, irq_priority, test_isr_0, 0, 0);
+
+	if (_sw_isr_table[irq_line_1].isr != test_isr_0) {
+		return false;
+	}
+	if (_sw_isr_table[irq_line_1].arg) {
+		return false;
+	}
+	if (_shared_irq_table[irq_line_1].client_num) {
+		return false;
+	}
+
+	arch_irq_connect_dynamic(irq_line_1, irq_priority, test_isr_1, (void *)1, 0);
+
+	if (_sw_isr_table[irq_line_1].isr != shared_isr) {
+		return false;
+	}
+	if (_sw_isr_table[irq_line_1].arg != &_shared_irq_table[irq_line_1]) {
+		return false;
+	}
+	if (_shared_irq_table[irq_line_1].client_num != 2) {
+		return false;
+	}
+
+	/* here it's mandatory that test_isr_0/NULL is the first client, while
+	 * test_isr_1/1 is the second client.
+	 */
+	if (!client_exists_at_index(test_isr_0, 0, irq_line_1, 0)) {
+		return 0;
+	}
+	if (!client_exists_at_index(test_isr_1, (void *)1, irq_line_1, 1)) {
+		return 0;
+	}
+
+	arch_irq_connect_dynamic(irq_line_2, irq_priority, test_isr_2, (void *)2, 0);
+
+	if (_sw_isr_table[irq_line_2].isr != test_isr_2) {
+		return false;
+	}
+	if (_sw_isr_table[irq_line_2].arg != (void *)2) {
+		return false;
+	}
+	if (_shared_irq_table[irq_line_2].client_num) {
+		return false;
+	}
+
+	irq_enable(irq_line_1);
+	irq_enable(irq_line_2);
+
+	trigger_irq(irq_line_1);
+	trigger_irq(irq_line_2);
+
+	/* wait 5ms before checking the results */
+	k_busy_wait(5000);
+
+	for (i = 0; i < TEST_VECTOR_SIZE; i++) {
+		if (test_vector[i] != result_vector[i]) {
+			return false;
+		}
+	}
+
+	irq_disable(irq_line_1);
+	irq_disable(irq_line_2);
+
+	reset_test_vector();
+
+	/* remove test_isr_0/NULL pair. After this statement we expect
+	 * irq_line_1 to be unshared.
+	 */
+	arch_irq_disconnect_dynamic(irq_line_1, irq_priority, test_isr_0, 0, 0);
+
+	if (_sw_isr_table[irq_line_1].isr != test_isr_1) {
+		return false;
+	}
+	if (_sw_isr_table[irq_line_1].arg != (void *)1) {
+		return false;
+	}
+	if (_shared_irq_table[irq_line_1].client_num) {
+		return false;
+	}
+
+	irq_enable(irq_line_1);
+	trigger_irq(irq_line_1);
+
+	/* wait 5ms before checking the results */
+	k_busy_wait(5000);
+
+	for (i = 0; i < TEST_VECTOR_SIZE; i++) {
+		if (i == 1) {
+			if (test_vector[i] != result_vector[i]) {
+				return false;
+			}
+			continue;
+		}
+
+		if (test_vector[i]) {
+			return false;
+		}
+	}
+
+	irq_disable(irq_line_1);
+
+	reset_test_vector();
+
+	/* pass */
+	return true;
+}
+
+ZTEST(interrupt_feature, test_shared_irq)
+{
+	/* TODO: make this test work on other architectures as well.
+	 * What are the equivalents of the SGIs on the other architectures?
+	 */
+#ifdef CONFIG_GIC
+	zassert_true(static_shared_irq_test(), "failed static shared irq test.");
+#endif /* CONFIG_GIC */
+
+#ifdef CONFIG_DYNAMIC_INTERRUPTS
+
+#ifdef CONFIG_GIC
+	zassert_true(reset_irq_table_state_test(), "failed reset table state test");
+#endif /* CONFIG_GIC */
+
+	zassert_true(dynamic_shared_irq_test(), "failed dynamic irq test.");
```

Comment:
Instead of using one single `ZTEST` and a set of `zassert_true`, use a `ZTEST` for each test.

Reply:
This is problematic because as far as I can tell the tests are executed in alphanumerical order. This causes problems since the mini tests expect `_sw_isr_table` and `_shared_isr_table` to be in a certain state. A workaround to this problem would be to change the test naming such that a certain order is followed but IMO this is not worth the trouble.